### PR TITLE
WIP Proximity sensor support

### PR DESCRIPTION
**Pull request by agx**
This adds proximity sensor support. The proximity of the vcnl4000 is single raw value. I'm marking as WIP since i'm not sure if this should be a different interface like compass or just go into the main interface?

### CODE REVIEW

**Comment by agx** [comment]
while working on #292 i noticed that drivers like the vcnl4000 that have both proximity and als currently have the issue that the same device can't have multiple IIO_SENSOR_PROXY_TYPE types. What would be the way forward here, making IIO_SENSOR_PROXY_TYPE a comma seperated list?

**Comment by hadess** [comment]
> while working on #292 i noticed that drivers like the vcnl4000 that have both proximity and als currently have the issue that the same device can't have multiple IIO_SENSOR_PROXY_TYPE types. What would be the way forward here, making IIO_SENSOR_PROXY_TYPE a comma seperated list?

That sounds good, yes.

As for whether the interface should be separate, or simply another one like light, I think it depends on whether we consider it a potential security risk. Pointers to what Android, Windows or iOS consider it to be would be useful in making a decision.

**Comment by agx** [comment]
> As for whether the interface should be separate, or simply another one like light, I think it depends on whether we consider it a potential security risk. Pointers to what Android, Windows or iOS consider it to be would be useful in making a decision.

android does not need extra permissions for the proximity sensor, there's no mention of permissions at https://developer.android.com/guide/topics/sensors/sensors_overview (see also https://stackoverflow.com/questions/20497087/manifest-xml-when-using-sensors), same seems to be true for iOS (https://developer.apple.com/documentation/uikit/uidevice/1620017-proximitymonitoringenabled?language=objc , https://developer.apple.com/documentation/uikit/uidevice)
so i'll drop the extra dbus interface. I'll also split the support for different sensor types types into an extra series since these are independent things.

File: src/drivers.h [hunk 1]
```diff
@@ -54,6 +59,10 @@ typedef struct {
 	gdouble heading;
 } CompassReadings;
 
+typedef struct {
+	int prox;
```
Comment (hadess) [inline src/drivers.h]: Do you want to export a simple "there's something close" boolean? Or the raw values from the sensor, or the values converted to meters, as the docs says we can?

From the docs:
```
What:           /sys/.../iio:deviceX/in_proximity_raw
What:           /sys/.../iio:deviceX/in_proximity_input
What:           /sys/.../iio:deviceX/in_proximityY_raw
KernelVersion:  3.4
Contact:        linux-iio@vger.kernel.org
Description:
                Proximity measurement indicating that some 
                object is near the sensor, usually by observing
                reflectivity of infrared or ultrasound emitted.
                Often these sensors are unit less and as such conversion
                to SI units is not possible. Higher proximity measurements
                indicate closer objects, and vice versa. Units after
                application of scale and offset are meters.
```

I think I would prefer if the D-Bus interface only ever sent a boolean. You'd either need to have a non-raw IIO proximity sensor (eg. it has scale and offset) or have those approximated by a udev property. Finally, we'd need to make a decision on what is considered close (maybe a couple of centimeters?) and make sure not to fluctuate too much when the position changes from ever so slightly above the threshold.

If your testing device has scale and offset, I'm fine with not having support for devices without those.

File: src/drv-iio-poll-proximity.c [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+	ReadingsUpdateFunc  callback_func;
+	gpointer            user_data;
+	GUdevDevice        *dev;
+	const char         *name;
+} DrvData;
+
+static DrvData *drv_data = NULL;
+
+static int
+sysfs_get_int (GUdevDevice *dev,
```
Comment (hadess) [inline src/drv-iio-poll-proximity.c]: `g_udev_device_get_sysfs_attr_as_int()`?

File: src/drv-iio-poll-proximity.c [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+
+static void
+iio_poll_proximity_set_polling (gboolean state)
+{
+	if (drv_data->timeout_id > 0 && state)
+		return;
+	if (drv_data->timeout_id == 0 && !state)
+		return;
+
+	if (drv_data->timeout_id) {
```
Comment (hadess) [inline src/drv-iio-poll-proximity.c]: This is new code, so `g_clear_handle_id()` (if you want to also make that change in other drivers, and bump the glib version required to 2.56, that'd be great)

File: README.md [hunk 4]
```diff
@@ -55,6 +55,8 @@ E: USEC_INITIALIZED=7750292
 You can now check whether a sensor is detected by running:
 ```
 gdbus introspect --system --dest net.hadess.SensorProxy --object-path /net/hadess/SensorProxy
+gdbus introspect --system --dest net.hadess.SensorProxy --object-path /net/hadess/SensorProxy/Compass
```
Comment (hadess) [inline README.md]: This isn't quite true, only the `geoclue` user is allowed to poke at that compass path. I'd leave that change out.

File: src/net.hadess.SensorProxy.xml [hunk 5]
```diff
@@ -174,4 +174,51 @@
...
+        clockwise order. That means North becomes 0 degree, East: 90 degrees,
+        South: 180 degrees, West: 270 degrees and so on. When unknown,
+        it's set to -1.0.
+    -->
+    <property name='ProximityProximity' type='i' access='read'/>
+
+    <!--
+       ClaimProximity:
+
+       To start receiving proximity heading updates from the proxy, the
```
Comment (hadess) [inline src/net.hadess.SensorProxy.xml]: Copy/paste error for this whole documentation section.

File: src/net.hadess.SensorProxy.xml [hunk 6]
```diff
@@ -174,4 +174,51 @@
...
+
+    <!--
+        ProximityHeading:
+
+        The proximity heading in degrees with respect to magnetic North direction, in
+        clockwise order. That means North becomes 0 degree, East: 90 degrees,
+        South: 180 degrees, West: 270 degrees and so on. When unknown,
+        it's set to -1.0.
+    -->
+    <property name='ProximityProximity' type='i' access='read'/>
```
Comment (hadess) [inline src/net.hadess.SensorProxy.xml]: We probably want that to be a bool.

**Review by hadess — CHANGES_REQUESTED** [review]

**Comment by hadess** [comment]
Moved to https://gitlab.freedesktop.org/hadess/iio-sensor-proxy/merge_requests/291

**Pull request closed without merging**